safeweb/http.go

```diff
@@ -0,0 +1,217 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package safeweb
+
+import (
+	"context"
+	crand "crypto/rand"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/gorilla/csrf"
+	"tailscale.com/tsnet"
+)
+
+// the default Content-Security-Policy header.
+// this policy:
+// - specifies the origin as the only valid source for all content types
+// - disallows inline javascript.
+// - disallows framing of the page
+// - disallows form submissions to other origins
+// - disallows base URIs from other origins
+// - disallows mixed content when serving over HTTPS
+// - disallows embedding of resources from other origins
+const defaultCSP string = `
+	default-src 'self';
+	script-src 'self';
+	frame-ancestors 'none';
+	form-action 'self';
+	base-uri 'self';
+	block-all-mixed-content;
+	object-src 'none'
+`
+
+type Config struct {
+	// Hostname is the name of the tsnet service that will be created to host
+	// the application.
+	Hostname string
+
+	// Listener is a net.Listener that the server will host the application on.
+	// Providing a listening will override the use of tsnet and the Hostname field.
+	Listener net.Listener
+
+	// BrowserMux is the HTTP handler for any routes in your application
+	// that should be served to browsers. These requests will be subject to CSRF
+	// protection and will have browser-specific headers set in response.
+	BrowserMux *http.ServeMux
+
+	// APIMux is the HTTP handler for any routes in your application that
+	// should be served to non-browser clients.
+	APIMux *http.ServeMux
+
+	// PermittedCrossOriginHosts specifies the Access-Control-Allow-Origin header sent in response to pre-flight OPTIONS requests.
+	PermittedCrossOriginHosts []string
+	// PermittedCrossOriginMethods specifies the Access-Control-Allow-Methods header sent in response to pre-flight OPTIONS requests.
+	PermittedCrossOriginMethods []string
+
+	// CSRFSecret is the secret used to sign CSRF tokens. It must be 32 bytes long.
```

Comment:
Do we care about the length, vs. that it be randomly generated?

Maybe also document that the caller should treat this as a sensitive value, though maybe "secret" is sufficient for that.

Reply:
Thoughts on specifying a `setec.Secret` here and create a pattern of hosting CSRF secrets at a standard key within the setec hierarchy for apps?

Reply:
I like the idea generally, but I feel like we should try to keep this package with as few dependencies as we can outside the standard library. My high-level notion is that if we decided to throw this out into its own repo for anyone to use, they should be able to do so without pulling in too much of the rest of Tailscale.

That said, I'm not opposed.

Reply:
In the interests of keeping things lightweight I'm going to leave it as a slice for the moment.